src/Razor/src/Microsoft.VisualStudio.LiveShare.Razor/Guest/ProjectSnapshotSynchronizationServiceFactory.cs

```diff
@@ -1,69 +1,40 @@
 using System.Threading;
 using System.Threading.Tasks;
-using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.Razor;
-using Microsoft.CodeAnalysis.Razor.ProjectSystem;
-using Microsoft.VisualStudio.LanguageServices;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
 using Microsoft.VisualStudio.LiveShare.Razor.Serialization;
 using Microsoft.VisualStudio.Threading;
 using Newtonsoft.Json;
 
 namespace Microsoft.VisualStudio.LiveShare.Razor.Guest;
 
 [ExportCollaborationService(typeof(ProjectSnapshotSynchronizationService), Scope = SessionScope.Guest)]
-internal class ProjectSnapshotSynchronizationServiceFactory : ICollaborationServiceFactory
+[method: ImportingConstructor]
+internal class ProjectSnapshotSynchronizationServiceFactory(
+    JoinableTaskContext joinableTaskContext,
+    IProjectSnapshotManagerAccessor projectManagerAccessor,
+    IErrorReporter errorReporter,
+    ProjectSnapshotManagerDispatcher dispatcher) : ICollaborationServiceFactory
 {
-    private readonly ProxyAccessor _proxyAccessor;
-    private readonly JoinableTaskContext _joinableTaskContext;
-    private readonly Workspace _workspace;
-
-    [ImportingConstructor]
-    public ProjectSnapshotSynchronizationServiceFactory(
-        ProxyAccessor proxyAccessor,
-        JoinableTaskContext joinableTaskContext,
-        [Import(typeof(VisualStudioWorkspace))] Workspace workspace)
-    {
-        if (proxyAccessor is null)
-        {
-            throw new ArgumentNullException(nameof(proxyAccessor));
-        }
-
-        if (joinableTaskContext is null)
-        {
-            throw new ArgumentNullException(nameof(joinableTaskContext));
-        }
-
-        if (workspace is null)
-        {
-            throw new ArgumentNullException(nameof(workspace));
-        }
-
-        _proxyAccessor = proxyAccessor;
-        _joinableTaskContext = joinableTaskContext;
-        _workspace = workspace;
-    }
-
     public async Task<ICollaborationService> CreateServiceAsync(CollaborationSession sessionContext, CancellationToken cancellationToken)
     {
         // This collaboration service depends on these serializers being immediately available so we need to register these now so the
         // guest project snapshot manager can retrieve the hosts project state.
         var serializer = (JsonSerializer)sessionContext.GetService(typeof(JsonSerializer));
         serializer.Converters.RegisterRazorLiveShareConverters();
 
-        var languageServices = _workspace.Services.GetLanguageServices(RazorLanguage.Name);
-        var projectManager = (ProjectSnapshotManagerBase)languageServices.GetRequiredService<ProjectSnapshotManager>();
-
         var projectSnapshotManagerProxy = await sessionContext.GetRemoteServiceAsync<IProjectSnapshotManagerProxy>(typeof(IProjectSnapshotManagerProxy).Name, cancellationToken);
         var synchronizationService = new ProjectSnapshotSynchronizationService(
-            _joinableTaskContext.Factory,
+            joinableTaskContext.Factory,
```

Comment:
IMO we should still capture this in a private field

Reply:
I go back in forth on this and there definitely isn't a hard-and-fast rule. In general, my rule of thumb has been to allow parameter capture for types with one method. In that case, I feel like the usage of the parameter is close enough to the declaration that I'm OK with it. What do you think?

Reply:
My gut reaction is to disagree, but it's not to the point that I'd block on it. In my mind it's based on the code I'm looking at. `_dispatcher` lets me know that the dispatcher is a locally owned field, while `dispatcher` lets me know it was either passed into the method or created/retrieved within the method. It's about quick knowledge of ownership and life expectancy of a variable

Reply:
I principal I agree, but in practice when I'm looking at a diff on GitHub I can't really tell that this is a small file with only one method.

Reply:
Totally agree with your gut feeling, but I come to a slightly different conclusion when the parameters are just data that are passed along. Primary constructors are the new var.

Reply:
Yes 100%. I'm sure in a year I'll have a picture hung on my wall of an underscore with "Remember the good times" inscibed on it, and be fully bought in to primary constructors

Reply:
The apocalypse will come when you started getting PR comments that say "nit: apply the 'Use Primary Constructor (and remove fields)' code action"

Reply:
I've also decided I'm going to have strong feelings about rewriting `x.ToArray()` into `[.. x]`. Got to have something to complain about!

Reply:
![GrumpyOldGIF](https://github.com/dotnet/razor/assets/475144/e9ee4e50-c7db-4426-8a80-82b08308c3f6)

Reply:
🌶️